### ISSUE TITLE
[Question]  can 'regctl export' only specific platforms?

### ISSUE THREAD

**Issue by lazywhite** [regclient/regclient#367]
### Question
It seems like ```regctl export alpine:3.16``` will export all platforms images,  can we export only some specific platforms like this?

```regctl export --platform  linux/amd64,linux/arm64 alpine:3.16```

**Comment by sudo-bmitch** [comment]
Hi @lazywhite. At present, that's not a feature of the export command. I'd be curious what you want the result to look like, since removing platforms from a multi-platform manifest leaves two options.

One is a sparse manifest, unchanged from upstream, but requests for the missing platforms will fail. There's an effort to add support for these to distribution/distribution, but most registries would reject them as an invalid manifest list.

The other option is to create a new manifest with only the two platforms. The result has a different digest making it difficult to compare to upstream (you have to compare each individual platform).

For the first option, there's a hidden flag to `regctl image copy`, which is hidden because of the lack of registry support and because I'm hoping to avoid support any issues it creates:

```
$ regctl image copy --platforms linux/amd64 --platforms linux/arm64 alpine:3.16 ocidir://test/alpine-sparse:3.16

$ regctl manifest get ocidir://test/alpine-sparse:3.16
Name:        ocidir://test/alpine-sparse:3.16
MediaType:   application/vnd.docker.distribution.manifest.list.v2+json
Digest:      sha256:1bd67c81e4ad4b8f4a5c1c914d7985336f130e5cefb3e323654fd09d6bcdbbe2
             
Manifests:   
             
  Name:      ocidir://test/alpine-sparse:3.16@sha256:0b29a7f4d42d6b5d6433ea91322903900e81b95d47d97d909a6e388e840f4f4a
  Digest:    sha256:0b29a7f4d42d6b5d6433ea91322903900e81b95d47d97d909a6e388e840f4f4a
  MediaType: application/vnd.docker.distribution.manifest.v2+json
  Platform:  linux/amd64
             
  Name:      ocidir://test/alpine-sparse:3.16@sha256:ed4d840b601f052e9d1c3bb843ad11e904d0265936c90c18e8e7bf6dc2f80d41
  Digest:    sha256:ed4d840b601f052e9d1c3bb843ad11e904d0265936c90c18e8e7bf6dc2f80d41
  MediaType: application/vnd.docker.distribution.manifest.v2+json
  Platform:  linux/arm/v6
             
  Name:      ocidir://test/alpine-sparse:3.16@sha256:32c3aacc36c2ceb104f43efbd7cf1c0732cf798e088b784b7c02cee27498b0a8
  Digest:    sha256:32c3aacc36c2ceb104f43efbd7cf1c0732cf798e088b784b7c02cee27498b0a8
  MediaType: application/vnd.docker.distribution.manifest.v2+json
  Platform:  linux/arm/v7
             
  Name:      ocidir://test/alpine-sparse:3.16@sha256:3b37be168bb81cf274793cebd596b4a023bdbca6f3b5bc3dfe8974e569b74feb
  Digest:    sha256:3b37be168bb81cf274793cebd596b4a023bdbca6f3b5bc3dfe8974e569b74feb
  MediaType: application/vnd.docker.distribution.manifest.v2+json
  Platform:  linux/arm64
             
  Name:      ocidir://test/alpine-sparse:3.16@sha256:abc178f562a2827f068bd0e60f48d998179dcd7130cc5036a287e1716536c306
  Digest:    sha256:abc178f562a2827f068bd0e60f48d998179dcd7130cc5036a287e1716536c306
  MediaType: application/vnd.docker.distribution.manifest.v2+json
  Platform:  linux/386
             
  Name:      ocidir://test/alpine-sparse:3.16@sha256:e6a8ba98c4be90c07b31be6e3315beeaa1fd5269ac215e581babef640a34ed0b
  Digest:    sha256:e6a8ba98c4be90c07b31be6e3315beeaa1fd5269ac215e581babef640a34ed0b
  MediaType: application/vnd.docker.distribution.manifest.v2+json
  Platform:  linux/ppc64le
             
  Name:      ocidir://test/alpine-sparse:3.16@sha256:1394c85a9984cb03d369381672a1767378b49d8701adfc6337258756052b2eeb
  Digest:    sha256:1394c85a9984cb03d369381672a1767378b49d8701adfc6337258756052b2eeb
  MediaType: application/vnd.docker.distribution.manifest.v2+json
  Platform:  linux/s390x

$ regctl manifest get ocidir://test/alpine-sparse:3.16 --platform linux/386
failed to pull platform specific digest: failed to open manifest: open test/alpine-sparse/blobs/sha256/abc178f562a2827f068bd0e60f48d998179dcd7130cc5036a287e1716536c306: no such file or directory

$ regctl manifest get ocidir://test/alpine-sparse:3.16 --platform linux/amd64
Name:        ocidir://test/alpine-sparse:3.16
MediaType:   application/vnd.docker.distribution.manifest.v2+json
Digest:      sha256:0b29a7f4d42d6b5d6433ea91322903900e81b95d47d97d909a6e388e840f4f4a
Total Size:  2.808MB
             
Config:      
  Digest:    sha256:dfd21b5a31f5ed07524f8e58d5e774efd13ce2e0a46d10852b7292d7394444c5
  MediaType: application/vnd.docker.container.image.v1+json
  Size:      1470B
             
Layers:      
             
  Digest:    sha256:ef5531b6e74e7865f5d142926b572d742ad47c153211de707903b4f5dd8d9e77
  MediaType: application/vnd.docker.image.rootfs.diff.tar.gzip
  Size:      2.808MB
```

For the second option using `regctl index create` command and some scripting will work:

```
$ regctl index create ocidir://test/alpine-partial:3.16 \
  --ref alpine@$(regctl image digest --platform linux/amd64 alpine:3.16) \
  --ref alpine@$(regctl image digest --platform linux/arm64 alpine:3.16)

$ regctl manifest get ocidir://test/alpine-partial:3.16
Name:        ocidir://test/alpine-partial:3.16
MediaType:   application/vnd.oci.image.index.v1+json
Digest:      sha256:404836783e079a51807d55452312c62c6b5094522b5f27a75d1085244a72272e
             
Manifests:   
             
  Name:      ocidir://test/alpine-partial:3.16@sha256:0b29a7f4d42d6b5d6433ea91322903900e81b95d47d97d909a6e388e840f4f4a
  Digest:    sha256:0b29a7f4d42d6b5d6433ea91322903900e81b95d47d97d909a6e388e840f4f4a
  MediaType: application/vnd.docker.distribution.manifest.v2+json
  Platform:  linux/amd64
             
  Name:      ocidir://test/alpine-partial:3.16@sha256:3b37be168bb81cf274793cebd596b4a023bdbca6f3b5bc3dfe8974e569b74feb
  Digest:    sha256:3b37be168bb81cf274793cebd596b4a023bdbca6f3b5bc3dfe8974e569b74feb
  MediaType: application/vnd.docker.distribution.manifest.v2+json
  Platform:  linux/arm64
```

The ocidir is the same as an export when you have multi-platform content (the shared docker export only works with a single platform). You can `tar -czf test/alpine-partial.tgz -C test/alpine-partial .` to get it packaged in a single file.

Some of these commands may request the `edge` version of images (or built off of the `main` branch). I'm working on getting another release out in the next week or so, because it's been way too long.

**Comment by lazywhite** [comment]
My company has a private registry, and we plan to run our products only on amd64 and arm64.
When we want to transfer images from docker.io to private registry, we want to preserve only that two platforms images

**Comment by sudo-bmitch** [comment]
If you need the digests to match, then you're best copying everything. If you don't care about the digests, want to prioritize disk space, and need more than one platform, then creating a new index is the current workaround.

I have it on my list to make workflows like this easier, so I'll leave this open until there's a better answer.

**Comment by sudo-bmitch** [comment]
I just merged #381 which provides a better way to create a manifest list from several platforms out of the source image. It's not directly in the export command, but you can create a manifest list in a local folder first using the ocidir scheme, and then tar or export that into a single file.

**Comment by lazywhite** [comment]
well done！ thx
